paddle/fluid/operators/fill_diagonal_tensor_op.cc

```diff
@@ -0,0 +1,284 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/fluid/operators/fill_diagonal_tensor_op.h"
+
+namespace paddle {
+namespace operators {
+
+// calculate the offset\new_dims\(strides of dim1/dim2)\matoffset
+void CalMatDims(framework::DDim out_dims, int dim1, int dim2, int *offset,
+                int *new_dims, int *strides, int *matoffset) {
+  int dimsum = 1, batchdim = 1;
+  int rank = out_dims.size();
+  int matoffidx = 0;
+  for (int i = rank - 1; i >= 0; i--) {
+    if (i == dim2) {
+      strides[0] = dimsum;
+    } else if (i == dim1) {
+      strides[1] = dimsum;
+    } else {
+      batchdim *= out_dims[i];
+      // matoffset calculate the offset position of the diagonal defined by dim1
+      // and dim2
+      // the first circle calculate the final free dimension
+      // and then calculate the front free dim one by one
+      if (matoffidx == 0) {
+        for (int j = 0; j < out_dims[i]; j++) {
+          matoffset[matoffidx] = dimsum * j;
+          matoffidx++;
+        }
+      } else {
+        auto size = matoffidx;
+        for (int j = 1; j < out_dims[i]; j++) {
+          for (int k = 0; k < size; k++) {
+            matoffset[matoffidx] = matoffset[k] + dimsum * j;
+            matoffidx++;
+          }
+        }
+      }
+    }
+    dimsum *= out_dims[i];
+  }
+
+  auto diagdim = dim1;
+  if (*offset >= 0) {
+    diagdim = std::min(out_dims[dim1], out_dims[dim2] - *offset);
+    *offset *= strides[0];
+  } else {
+    diagdim = std::min(out_dims[dim1] + *offset, out_dims[dim2]);
+    *offset *= -strides[1];
+  }
+  new_dims[0] = batchdim;
+  new_dims[1] = diagdim;
+  return;
+}
+
+class FillDiagonalTensorOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddComment(R"DOC(Fill replace operator
+                Fill the diagonal matrix of an tensor with `src_tensor` Tensor.
+                )DOC");
+    AddInput("X", "(Tensor) The input tensor.");
+    AddInput("src_tensor", "(Tensor) The input tensor to filled in.");
+    AddOutput("Out",
+              "Tensor, the output tensor, with the same shape and data type "
+              "as input(x)");
+    AddAttr<int>("dim1", "the first dim to figure out the diagonal")
+        .SetDefault(-2);
+    AddAttr<int>("dim2", "the second dim to figure out the diagonal")
+        .SetDefault(-1);
+    AddAttr<int>("offset",
+                 "offset of diagonal, zero means no offset, positive means "
+                 "offset to up-right corner; negtive means offset to "
+                 "bottom-left corner")
+        .SetDefault(0);
+    AddAttr<int>("matrows",
```

Comment:
used for backward memory allocate

Reply:
I know the purpose that you add this attribute. I mean, this attribute is not necessary, because it can be calculated by other informations (i.e., the dimensions of the input tensors and the values of other attributes). It is not **orthogonal** with other attributes.

Reply:
done